### DISCUSSION TITLE
Tool Return Type

### DISCUSSION THREAD

**Discussion by Ga68** [VRSEN/agency-swarm#145] [Q&A] [answered]
Does a tool NEED to return a string? Is it a best practice? If it doesn't return a string, will it get converted with a json-dump or maybe the str or repr Dunder?

**Comment by nicko-ai** [comment, marked as answer]
Hey @Ga68,
  Tools do NOT need to return a string, but whatever they return will be converted to a string before being sent to OpenAI.

  Key Findings:

  1. Automatic String Conversion: In /Users/nick/Areas/agency-swarm/agency_swarm/threads/thread.py:775-777, there's explicit code that converts non-string outputs to strings:
```
  for to_ in tool_outputs:
      if not isinstance(to_["output"], str):
          to_["output"] = str(to_["output"])
```
  2. Return Type Flexibility: Tools can return:
    - Strings (most common)
    - Any Python object (will be converted via str())
    - Coroutines (for async operations)
  3. Best Practices:
    - Return strings when possible for clarity and control over formatting
    - If returning complex objects, implement a meaningful __str__() method
    - The conversion uses Python's built-in str() function, not json.dumps() or repr()
  4. Examples from the codebase:
    - FileReader returns a formatted string with line numbers
    - ClickElement returns descriptive result strings
    - SendMessage returns the completion response (string)

So to directly answer the question: No, tools don't NEED to return strings, but it's a best practice since non-string returns will be converted using str() anyway. You have full control over the output format by returning a properly formatted string rather than relying on the default string conversion.
